flang/lib/Lower/IO.cpp

```diff
@@ -526,8 +542,12 @@ static mlir::Value createIoRuntimeCallForItem(mlir::Location loc,
                    itemTy.cast<mlir::IntegerType>().getWidth() / 8)));
     }
   }
-  return builder.create<fir::CallOp>(loc, inputFunc, inputFuncArgs)
-      .getResult(0);
+  auto call = builder.create<fir::CallOp>(loc, inputFunc, inputFuncArgs);
+  auto itemAddr = fir::getBase(item);
+  auto itemTy = fir::unwrapPassByRefType(itemAddr.getType());
+  if (itemTy.isa<fir::LogicalType>())
+    boolRefToLogical(loc, builder, itemAddr);
```

Comment:
Seems like another approach might be to normalize the LOGICAL in memory to be `0` or `1` before the call. It sounds like the runtime doesn't use/know about the more significant bytes.